lib/orchestrator.js

```diff
@@ -0,0 +1,354 @@
+          return new Map({
+            status: 'ok',
+            history: 'The document passed the third party checker.'
+          });
+        }
+        else {
+          return new Map({
+            history: 'The document contains non-authorized resources',
+            status: 'failure',
+            errors: extResources
+          });
+        }
+      }).catch(function (error) {
+        return new Map({
+          status: 'error',
+          errors: List.of(error.toString()),
+          history:
+            'An error occurred while running the Third Party Resources Checker.'
+        });
+      })
+  });
+};
+
+Orchestrator.prototype.runPublisher = function (metadata) {
+  var pubsystemService = new JsonHttpService(
+    global.W3C_PUBSYSTEM_URL,
+    global.USERNAME,
+    global.PASSWORD
+  );
+
+  return new Map({
+    name: 'publish',
+    promise: new Publisher(pubsystemService).publish(metadata)
+      .then(function (errors) {
+        if (errors.isEmpty()) {
+          return new Map({
+            status: 'ok'
+          });
+        }
+        else {
+          return new Map({
+            status: 'failure',
+            errors: errors,
+            history: 'The document could not be published: ' +
+              errors.map(function (e) { return e.message; })
+          });
+        }
+      }).catch(function (error) {
+        return new Map({
+          status: 'error',
+          errors: List.of(error.toString()),
+          history: 'The document could not be published: ' + error.message
+        });
+      })
+  });
+};
+
+Orchestrator.prototype.runTrInstaller = function (thisVersion, tempLocation) {
+  var finalTRpath = thisVersion.replace('http://www.w3.org', '');
+  return new Map({
+    name: 'tr-install',
+    promise: trInstaller(tempLocation, finalTRpath).then(function () {
+      return new Map({ status: 'ok' });
+    }).catch(function (error) {
+      return new Map({
+        status: 'error',
+        errors: List.of(error.toString()),
+        history: 'An error occurred while installing the document in TR'
+      });
+    })
+  });
+};
+
+Orchestrator.prototype.runShortlink = function (thisVersion) {
+  return new Map({
+    name: 'update-tr-shortlink',
+    promise: updateTrShortlink(thisVersion).then(function () {
+      return new Map({ status: 'ok' });
+    }).catch(function (error) {
+      return new Map({
+        status: 'error',
+        errors: List.of(error.toString()),
+        history: 'An error occurred while updating the shortlink.'
+      });
+    })
+  });
+};
+
+/**
+ * Run the next step of the orchestration, for a given state.
+ *
+ * @param {RequestState} state - A state for a specific publication request
+ * @returns {List.<Promise.<RequestState>>} A list of promised states that are fulfilled at different stages of each step
+ */
+Orchestrator.prototype.next = function (state) {
+  if (state.hasJobStarted('retrieve-resources')) {
+    return this.runStep(
```

Comment:
instead of repeating this.runStep(…)(state), put the step to be run in a variable, and call runStep() only once?